File: thermion_dart/native/src/scene/AnimationManager.cpp
#include <memory>
#include <stack>
#include <unordered_set>
#include <vector>

#include <filament/Engine.h>
#include <filament/TransformManager.h>
#include <filament/RenderableManager.h>

#include <gltfio/Animator.h>

#include <utils/Panic.h>

#include "Log.hpp"

#include "scene/AnimationManager.hpp"

#include "components/GltfAnimationComponentManager.hpp"
#include "components/MorphAnimationComponentManager.hpp"
#include "components/BoneAnimationComponentManager.hpp"

#include "scene/SceneAsset.hpp"
#include "scene/GltfSceneAssetInstance.hpp"

namespace thermion
{

    using namespace filament;
    using namespace utils;

    AnimationManager::AnimationManager(Engine *engine) : mEngine(engine)
    {
        auto &transformManager = mEngine->getTransformManager();
        auto &renderableManager = mEngine->getRenderableManager();
        auto &em = EntityManager::get();
        mGltfAnimationComponentManager = std::make_unique<GltfAnimationComponentManager>(em, transformManager, renderableManager);
        mMorphAnimationComponentManager = std::make_unique<MorphAnimationComponentManager>(em, transformManager, renderableManager);
        mBoneAnimationComponentManager = std::make_unique<BoneAnimationComponentManager>(em, transformManager, renderableManager);
    }

    AnimationManager::~AnimationManager()
    {
    }

    bool AnimationManager::setMorphAnimationBuffer(
        utils::Entity entity,
        const float *const morphData,
        const uint32_t *const morphIndices,
        int numMorphTargets,
        int numFrames,
        float frameLengthInMs)
    {

        std::lock_guard lock(mMutex);

        if (!mMorphAnimationComponentManager->hasComponent(entity))
        {
            mMorphAnimationComponentManager->addAnimationComponent(entity);
        }

        auto animationComponentInstance = mMorphAnimationComponentManager->getInstance(entity);
        auto &animationComponent = mMorphAnimationComponentManager->elementAt<0>(animationComponentInstance);
        auto &morphAnimations = animationComponent.animations;

        MorphAnimation morphAnimation;

        morphAnimation.frameData.clear();
        morphAnimation.frameData.insert(
            morphAnimation.frameData.begin(),
            morphData,
            morphData + (numFrames * numMorphTargets));
        morphAnimation.frameLengthInMs = frameLengthInMs;
        morphAnimation.morphIndices.resize(numMorphTargets);
        for (int i = 0; i < numMorphTargets; i++)
        {
            morphAnimation.morphIndices[i] = morphIndices[i];
        }
        morphAnimation.durationInSecs = (frameLengthInMs * numFrames) / 1000.0f;

        morphAnimation.startTimeInNanos = 0; // Will be set when first update() is called
        morphAnimation.lengthInFrames = numFrames;

        morphAnimations.emplace_back(morphAnimation);

        return true;
    }

    void AnimationManager::clearMorphAnimationBuffer(
        utils::Entity entity)
    {
        std::lock_guard lock(mMutex);

        auto animationComponentInstance = mMorphAnimationComponentManager->getInstance(entity);
        auto &animationComponent = mMorphAnimationComponentManager->elementAt<0>(animationComponentInstance);
        auto &morphAnimations = animationComponent.animations;
        morphAnimations.clear();
    }

    void AnimationManager::resetToRestPose(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);

        auto filamentInstance = instance->getInstance();
        auto skinCount = filamentInstance->getSkinCount();

        TransformManager &transformManager = mEngine->getTransformManager();

        //
        // To reset the skeleton to its rest pose, we could just call animator->resetBoneMatrices(),
        // which sets all bone matrices to the identity matrix. However, any subsequent calls to animator->updateBoneMatrices()
        // may result in unexpected poses, because that method uses each bone's transform to calculate
        // the bone matrices (and resetBoneMatrices does not affect this transform).
        // To "fully" reset the bone, we need to set its local transform (i.e. relative to its parent)
        // to its original orientation in rest pose.
        //
        // This can be calculated as:
        //
        //   auto rest = inverse(parentTransformInModelSpace) * bindMatrix
        //
        // (where bindMatrix is the inverse of the inverseBindMatrix).
        //
        // The only requirement is that parent bone transforms are reset before child bone transforms.
        // glTF/Filament does not guarantee that parent bones are listed before child bones under a FilamentInstance.
        // We ensure that parents are reset before children by:
        // - pushing all bones onto a stack
        // - iterate over the stack
        //      - look at the bone at the top of the stack
        //      - if the bone already been reset, pop and continue iterating over the stack
        //      - otherwise
        //          - if the bone has a parent that has not been reset, push the parent to the top of the stack and continue iterating
        //          - otherwise
        //              - pop the bone, reset its transform and mark it as completed
        for (int skinIndex = 0; skinIndex < skinCount; skinIndex++)
        {
            std::unordered_set<Entity, Entity::Hasher> joints;
            std::unordered_set<Entity, Entity::Hasher> completed;
            std::stack<Entity> stack;

            auto transforms = getBoneRestTranforms(instance, skinIndex);

            for (int i = 0; i < filamentInstance->getJointCountAt(skinIndex); i++)
            {
                auto restTransform = transforms[i];
                const auto &joint = filamentInstance->getJointsAt(skinIndex)[i];
                auto transformInstance = transformManager.getInstance(joint);
                transformManager.setTransform(transformInstance, restTransform);
            }
        }
        filamentInstance->getAnimator()->updateBoneMatrices();
        return;
    }

    std::vector<math::mat4f> AnimationManager::getBoneRestTranforms(GltfSceneAssetInstance *instance, int skinIndex)
    {

        std::vector<math::mat4f> transforms;

        auto filamentInstance = instance->getInstance();
        auto skinCount = filamentInstance->getSkinCount();

        TransformManager &transformManager = mEngine->getTransformManager();

        transforms.resize(filamentInstance->getJointCountAt(skinIndex));

        //
        // To reset the skeleton to its rest pose, we could just call animator->resetBoneMatrices(),
        // which sets all bone matrices to the identity matrix. However, any subsequent calls to animator->updateBoneMatrices()
        // may result in unexpected poses, because that method uses each bone's transform to calculate
        // the bone matrices (and resetBoneMatrices does not affect this transform).
        // To "fully" reset the bone, we need to set its local transform (i.e. relative to its parent)
        // to its original orientation in rest pose.
        //
        // This can be calculated as:
        //
        //   auto rest = inverse(parentTransformInModelSpace) * bindMatrix
        //
        // (where bindMatrix is the inverse of the inverseBindMatrix).
        //
        // The only requirement is that parent bone transforms are reset before child bone transforms.
        // glTF/Filament does not guarantee that parent bones are listed before child bones under a FilamentInstance.
        // We ensure that parents are reset before children by:
        // - pushing all bones onto a stack
        // - iterate over the stack
        //      - look at the bone at the top of the stack
        //      - if the bone already been reset, pop and continue iterating over the stack
        //      - otherwise
        //          - if the bone has a parent that has not been reset, push the parent to the top of the stack and continue iterating
        //          - otherwise
        //              - pop the bone, reset its transform and mark it as completed
        std::vector<Entity> joints;
        std::unordered_set<Entity, Entity::Hasher> completed;
        std::stack<Entity> stack;

        for (int i = 0; i < filamentInstance->getJointCountAt(skinIndex); i++)
        {
            const auto &joint = filamentInstance->getJointsAt(skinIndex)[i];
            joints.push_back(joint);
            stack.push(joint);
        }

        while (!stack.empty())
        {
            const auto &joint = stack.top();

            // if we've already handled this node previously (e.g. when we encountered it as a parent), then skip
            if (completed.find(joint) != completed.end())
            {
                stack.pop();
                continue;
            }

            const auto transformInstance = transformManager.getInstance(joint);
            auto parent = transformManager.getParent(transformInstance);

            // we need to handle parent joints before handling their children
            // therefore, if this joint has a parent that hasn't been handled yet,
            // push the parent to the top of the stack and start the loop again
            const auto &jointIter = std::find(joints.begin(), joints.end(), joint);
            auto parentIter = std::find(joints.begin(), joints.end(), parent);

            if (parentIter != joints.end() && completed.find(parent) == completed.end())
            {
                stack.push(parent);
                continue;
            }

            // otherwise let's get the inverse bind matrix for the joint
            math::mat4f inverseBindMatrix;
            bool found = false;
            for (int i = 0; i < filamentInstance->getJointCountAt(skinIndex); i++)
            {
                if (filamentInstance->getJointsAt(skinIndex)[i] == joint)
                {
                    inverseBindMatrix = filamentInstance->getInverseBindMatricesAt(skinIndex)[i];
                    found = true;
                    break;
                }
            }
            ASSERT_PRECONDITION(found, "Failed to find inverse bind matrix for joint %d", joint);

            // now we need to ascend back up the hierarchy to calculate the modelSpaceTransform
            math::mat4f modelSpaceTransform;
            while (parentIter != joints.end())
            {
                const auto transformInstance = transformManager.getInstance(parent);
                const auto parentIndex = distance(joints.begin(), parentIter);
                const auto transform = transforms[parentIndex];
                modelSpaceTransform = transform * modelSpaceTransform;
                parent = transformManager.getParent(transformInstance);
                parentIter = std::find(joints.begin(), joints.end(), parent);
            }

            const auto bindMatrix = inverse(inverseBindMatrix);

            const auto inverseModelSpaceTransform = inverse(modelSpaceTransform);

            const auto jointIndex = distance(joints.begin(), jointIter);
            transforms[jointIndex] = inverseModelSpaceTransform * bindMatrix;
            completed.insert(joint);
            stack.pop();
        }
        return transforms;
    }

    void AnimationManager::updateBoneMatrices(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);
        instance->getInstance()->getAnimator()->updateBoneMatrices();
    }

    bool AnimationManager::addBoneAnimation(GltfSceneAssetInstance *instance,
                                            int skinIndex,
                                            int boneIndex,
                                            const float *const frameData,
                                            int numFrames,
                                            float frameLengthInMs,
                                            float fadeOutInSecs,
                                            float fadeInInSecs,
                                            float maxDelta,
                                            bool loop)
    {
        std::lock_guard lock(mMutex);

        BoneAnimation animation;
        animation.boneIndex = boneIndex;
        animation.frameData.clear();

        const auto &inverseBindMatrix = instance->getInstance()->getInverseBindMatricesAt(skinIndex)[boneIndex];
        for (int i = 0; i < numFrames; i++)
        {
            math::mat4f frame(
                frameData[i * 16],
                frameData[(i * 16) + 1],
                frameData[(i * 16) + 2],
                frameData[(i * 16) + 3],
                frameData[(i * 16) + 4],
                frameData[(i * 16) + 5],
                frameData[(i * 16) + 6],
                frameData[(i * 16) + 7],
                frameData[(i * 16) + 8],
                frameData[(i * 16) + 9],
                frameData[(i * 16) + 10],
                frameData[(i * 16) + 11],
                frameData[(i * 16) + 12],
                frameData[(i * 16) + 13],
                frameData[(i * 16) + 14],
                frameData[(i * 16) + 15]);

            animation.frameData.push_back(frame);
        }

        animation.frameLengthInMs = frameLengthInMs;
        animation.startTimeInNanos = 0; // Will be set when first update() is called
        animation.startOffset = 0.0f;
        animation.reverse = false;
        animation.loop = loop;
        animation.durationInSecs = (frameLengthInMs * numFrames) / 1000.0f;
        animation.lengthInFrames = numFrames;
        animation.frameLengthInMs = frameLengthInMs;
        animation.fadeOutInSecs = fadeOutInSecs;
        animation.fadeInInSecs = fadeInInSecs;
        animation.maxDelta = maxDelta;
        animation.skinIndex = skinIndex;
        if (!mBoneAnimationComponentManager->hasComponent(instance->getInstance()->getRoot()))
        {
            Log("ERROR: specified entity is not animatable (has no animation component attached).");
            return false;
        }
        auto animationComponentInstance = mBoneAnimationComponentManager->getInstance(instance->getInstance()->getRoot());

        auto &animationComponent = mBoneAnimationComponentManager->elementAt<0>(animationComponentInstance);
        auto &boneAnimations = animationComponent.animations;

        boneAnimations.emplace_back(animation);

        return true;
    }

    void AnimationManager::playGltfAnimation(GltfSceneAssetInstance *instance, int index, bool loop, bool reverse, bool replaceActive, float crossfade, float startOffset, float speed)
    {
        std::lock_guard lock(mMutex);

        if (index < 0)
        {
            Log("ERROR: glTF animation index must be greater than zero.");
            return;
        }

        mGltfAnimationComponentManager->addGltfAnimation(instance->getInstance(), index, loop, reverse, replaceActive, crossfade, startOffset, speed);
    }

    void AnimationManager::stopGltfAnimation(GltfSceneAssetInstance *instance, int index)
    {
        std::lock_guard lock(mMutex);
        auto animationComponentInstance = mGltfAnimationComponentManager->getInstance(instance->getEntity());
        auto &animationComponent = mGltfAnimationComponentManager->elementAt<0>(animationComponentInstance);

        auto erased = std::remove_if(animationComponent.animations.begin(),
                                     animationComponent.animations.end(),
                                     [=](GltfAnimation &anim)
                                     { return anim.index == index; });
        animationComponent.animations.erase(erased,
                                            animationComponent.animations.end());
        return;
    }

    void AnimationManager::setMorphTargetWeights(utils::Entity entity, const float *const weights, const int count)
    {
        std::lock_guard lock(mMutex);
        RenderableManager &rm = mEngine->getRenderableManager();
        auto renderableInstance = rm.getInstance(entity);

        rm.setMorphWeights(
            renderableInstance,
            weights,
            count);
    }

    void AnimationManager::setGltfAnimationTime(GltfSceneAssetInstance *instance, int animationIndex, float timeInSeconds)
    {
        std::lock_guard lock(mMutex);
        instance->getInstance()->getAnimator()->applyAnimation(animationIndex, timeInSeconds);
        instance->getInstance()->getAnimator()->updateBoneMatrices();
        return;
    }

    float AnimationManager::getGltfAnimationDuration(GltfSceneAssetInstance *instance, int animationIndex)
    {
        return instance->getInstance()->getAnimator()->getAnimationDuration(animationIndex);
    }

    std::vector<std::string> AnimationManager::getGltfAnimationNames(GltfSceneAssetInstance *instance)
    {
        std::vector<std::string> names;

        size_t count = instance->getInstance()->getAnimator()->getAnimationCount();

        for (size_t i = 0; i < count; i++)
        {
            names.push_back(instance->getInstance()->getAnimator()->getAnimationName(i));
        }
        return names;
    }

    std::vector<std::string> AnimationManager::getMorphTargetNames(GltfSceneAsset *asset, EntityId childEntity)
    {
        std::vector<std::string> names;

        auto filamentAsset = asset->getAsset();

        const utils::Entity targetEntity = utils::Entity::import(childEntity);

        size_t count = filamentAsset->getMorphTargetCountAt(targetEntity);
        for (int j = 0; j < count; j++)
        {
            const char *morphName = filamentAsset->getMorphTargetNameAt(targetEntity, j);
            names.push_back(morphName);
        }
        return names;
    }

    void AnimationManager::update(uint64_t frameTimeInNanos)
    {
        std::lock_guard lock(mMutex);
        mGltfAnimationComponentManager->update(frameTimeInNanos);
        mMorphAnimationComponentManager->update(frameTimeInNanos);
        mBoneAnimationComponentManager->update(frameTimeInNanos);
        mLastUpdateTime = frameTimeInNanos;
    }

    math::mat4f AnimationManager::getInverseBindMatrix(GltfSceneAssetInstance *instance, int skinIndex, int boneIndex)
    {
        return instance->getInstance()->getInverseBindMatricesAt(skinIndex)[boneIndex];
    }

    bool AnimationManager::addGltfAnimationComponent(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);
        mGltfAnimationComponentManager->addAnimationComponent(instance->getInstance());
        TRACE("Added glTF animation component");
        return true;
    }

    void AnimationManager::removeGltfAnimationComponent(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);
        mGltfAnimationComponentManager->removeAnimationComponent(instance->getInstance());
        TRACE("Removed glTF animation component");
    }

    bool AnimationManager::addBoneAnimationComponent(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);
        mBoneAnimationComponentManager->addAnimationComponent(instance->getInstance());
        TRACE("Added bone animation component");
        return true;
    }

    void AnimationManager::removeBoneAnimationComponent(GltfSceneAssetInstance *instance)
    {
        std::lock_guard lock(mMutex);
        mBoneAnimationComponentManager->removeAnimationComponent(instance->getInstance());
        TRACE("Removed bone animation component");
    }

    bool AnimationManager::addMorphAnimationComponent(utils::Entity entity)
    {
        std::lock_guard lock(mMutex);
        mMorphAnimationComponentManager->addAnimationComponent(entity);
        TRACE("Added morph animation component");
        return true;
    }

    void AnimationManager::removeMorphAnimationComponent(utils::Entity entity)
    {
        std::lock_guard lock(mMutex);
        mMorphAnimationComponentManager->removeAnimationComponent(entity);
        TRACE("Removed morph animation component");
    }

}

File: thermion_dart/lib/src/filament/src/interface/animation_manager.dart
import 'package:thermion_dart/src/filament/src/interface/native_handle.dart';
import 'package:thermion_dart/thermion_dart.dart';

// This is a thin wrapper over the (C++) AnimationManager, which manages glTF
// animations (i.e. animations embedded in glTF assets), as well as
// bone/skeletal animations and/or morph target animations (which users would
// construct manually).
//
// On every frame, the C++ AnimationManager iterates over all animation
// components and performs any updates necessary to advance animations
// that have not yet completed:
// - for glTF animation components, AnimationManager calls the Animator
//   methods on the underlying FilamentInstance
// - for morph target components, AnimationManager calls
//   RenderableManager::setMorphWeights
// - for bone animation components, AnimationManager calls setTransform directly
//   on the bone entities
//
// The (Dart) AnimationManager class is a thin wrapper over the C++ class.
abstract class AnimationManager<T> extends NativeHandle<T> {
  // Adds a glTF animation component to [asset]. These are added at the asset
  // level (rather than entity) because the underlying animator drives the
  // entire FilamentInstance, rather than individual entities.
  // Returns true if successful, false otherwise
  Future<bool> addGltfAnimationComponent(ThermionAsset asset);

  // Removes the glTF animation component from [asset].
  // Returns true if successful, false otherwise.
  Future<bool> removeGltfAnimationComponent(ThermionAsset asset);

  // Adds a morph animation component to [entity].
  // These are added at the entity level (rather than the asset level) because
  // morph targets
  void addMorphAnimationComponent(ThermionEntity entity);

  // Removes the morph animation component from [entity].
  void removeMorphAnimationComponent(ThermionEntity entity);

  // Adds a bone animation component to the specified asset.
  Future<bool> addBoneAnimationComponent(ThermionAsset asset);

  // Removes the bone animation component from the specified asset.
  Future<bool> removeBoneAnimationComponent(ThermionAsset asset);

  // Plays a glTF animation with the specified parameters.
  //
  // [asset] The asset containing the animation
  // [index] The animation index to play
  // [loop] Whether to loop the animation
  // [reverse] Whether to play the animation in reverse
  // [replaceActive] Whether to replace any currently active animation
  // [crossfade] The crossfade duration in seconds when transitioning
  // [startOffset] The offset into the animation to start from (0.0 to 1.0)
  // [speed] The playback speed multiplier (1.0 = normal speed)
  // Returns true if successful, false otherwise
  bool playGltfAnimation(
    ThermionAsset asset,
    int index, {
    bool loop = false,
    bool reverse = false,
    bool replaceActive = true,
    double crossfade = 0.0,
    double startOffset = 0.0,
    double speed = 1.0,
  });

  // Stops a currently playing glTF animation.
  //
  // [asset] The asset containing the animation
  // [index] The animation index to stop
  // Returns true if successful, false otherwise
  bool stopGltfAnimation(ThermionAsset asset, int index);

  // Sets the current frame for a glTF animation.
  //
  // [asset] The asset containing the animation
  // [animationIndex] The animation index
  // [timeInSeconds] The time position in the animation in seconds
  // Returns true if successful, false otherwise
  Future<void> setGltfAnimationTime(ThermionAsset asset, int animationIndex, double timeInSeconds);

  // Gets the duration of a glTF animation in seconds.
  //
  // [asset] The asset containing the animation
  // [animationIndex] The animation index
  // Returns the duration in seconds, or 0.0 if not found
  double getGltfAnimationDuration(ThermionAsset asset, int animationIndex);

  // Gets the number of glTF animations available in the asset.
  //
  // [asset] The asset to query
  // Returns the number of animations
  int getGltfAnimationCount(ThermionAsset asset);

  // Gets the name of a glTF animation.
  //
  // [asset] The asset containing the animation
  // [index] The animation index
  // Returns the animation name, or null if not found
  String? getGltfAnimationName(ThermionAsset asset, int index);

  // Sets up a morph animation with keyframe data.
  //
  // [entityId] The entity to animate
  // [morphData] List of morph target weight values for each frame
  // [morphIndices] Indices of morph targets to animate
  // [numMorphTargets] Number of morph targets being animated
  // [numFrames] Number of animation frames
  // [frameLengthInMs] Length of each frame in milliseconds
  // Returns true if successful, false otherwise
  bool setMorphAnimation(
    ThermionEntity entityId,
    List<double> morphData,
    List<int> morphIndices,
    int numMorphTargets,
    int numFrames,
    double frameLengthInMs,
  );

  // Clears all morph animation data for an entity.
  //
  // [entityId] The entity to clear animation for
  // Returns true if successful, false otherwise
  bool clearMorphAnimation(ThermionEntity entityId);

  // Gets the number of morph target names for an entity.
  //
  // [asset] The asset containing the entity
  // [entityId] The entity to query
  // Returns the number of morph target names
  int getMorphTargetNameCount(ThermionAsset asset, ThermionEntity entityId);

  // Gets the name of a morph target at the specified index.
  //
  // [asset] The asset containing the entity
  // [entityId] The entity to query
  // [index] The morph target index
  // Returns the morph target name, or null if not found
  String? getMorphTargetName(ThermionAsset asset, ThermionEntity entityId, int index);

  // Adds bone animation data with keyframes.
  //
  // [asset] The asset containing the bones
  // [skinIndex] The skin index containing the bones
  // [boneIndex] The bone index to animate
  // [frameData] Transformation data for each frame
  // [numFrames] Number of animation frames
  // [frameLengthInMs] Length of each frame in milliseconds
  // [fadeOutInSecs] Fade out duration in seconds
  // [fadeInInSecs] Fade in duration in seconds
  // [maxDelta] Maximum transformation delta per frame
  // Returns true if successful, false otherwise
  Future<bool> addBoneAnimation(
    ThermionAsset asset,
    int skinIndex,
    int boneIndex,
    List<double> frameData,
    int numFrames,
    double frameLengthInMs, {
    double fadeOutInSecs = 0.0,
    double fadeInInSecs = 0.0,
    double maxDelta = 0.1,
    bool loop,
  });

  // Gets the rest pose local transforms for all bones in a skin.
  //
  // [asset] The asset containing the bones
  // [skinIndex] The skin index
  // Returns a list of 4x4 transformation matrices (16 floats per bone)
  Future<List<double>> getRestLocalTransforms(ThermionAsset asset, int skinIndex);

  // Gets the inverse bind matrix for a specific bone.
  //
  // [asset] The asset containing the bones [skinIndex] The skin index
  // [boneIndex] The bone index Returns the 4x4 inverse bind matrix (16
  // floats), or empty list if not found
  Future<List<double>> getInverseBindMatrix(ThermionAsset asset, int skinIndex, int boneIndex);

  // Calls [updateBoneMatrices] on the Animator instance for the
  // FilamentInstance (which uses the local bone transform and the inverse bind
  // matrix to set the bone matrix). [asset] The asset to update. This must be
  // an instance. Returns true if successful, false otherwise
  Future<bool> updateBoneMatrices(ThermionAsset asset);

  // Resets [asset] to its rest pose.
  Future resetToRestPose(ThermionAsset asset);

  // Updates the animation system with the current frame time.
  // [frameTimeInNanos] Frame time in nanoseconds
  Future update(int frameTimeInNanos);
}
